### ISSUE TITLE
idea: add `FLUX` environment variable that holds "closest enclosing jobid"

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#6474]
This idea was brought up by @trws on slack and in a project meeting.

Flux currently doesn't have a consistent way to determine the "nearest" enclosing jobid. The cases are somewhat delineated in #3817, though the information there may be outdated (i.e. there does now exist a [flux_job_timeleft(3)](https://flux-framework.readthedocs.io/projects/flux-core/en/latest/man3/flux_job_timeleft.html) function). While it makes sense that `FLUX_JOB_ID` is set in the environment of tasks launched by `flux run` and `flux submit` but not in the environment of the initial program in `flux alloc` and `flux batch`, this will likely continue to cause confusion and annoyance for users.

One idea put forward by @trws is to add another `FLUX_` jobid variable that is always set by the job shell which is _not_ cleared. (Please correct me if I'm mistaken). This variable would leak through to initial programs, which would then be able to use this variable to determine the jobid of their parent instance (if there was a jobid associated) - equivalent to, but more straightforward than using `flux getattr jobid`. It would _also_ be available in `flux run` and `flux submit` where it would be the same as `FLUX_JOB_ID`. Comparing the two variables would allow users to easily determine if they are in an initial program environment or within a job. Lack of this new environment variable would indicate that there is no enclosing job, i.e. the current process is not within an instance, or the enclosing instance is not itself a job.

Edit: if we enable this feature, that may allow us to close #3817.

**Comment by chu11** [comment]
a random idea I thought of.  if we want to avoid spreading too many environment variables, could we support a new command like hypothetically `flux job whoami`?  (i.e. like `flux job last`).

**Comment by grondo** [comment]
That's a pretty good idea. I wonder if that would satisfy @vsoch and @trws (I think it probably would, but they should weigh in just as well)?

**Comment by garlick** [comment]
Would `flux job whoami` just be an alias for `flux getattr jobid` then?

**Comment by grondo** [comment]
I think if `FLUX_JOB_ID` is set, it would return that, otherwise `flux getattr jobid`. If that failed, then it would return empty?

**Comment by vsoch** [comment]
An environment variable or flux attribute (each of which works consistently across cases) would be great. Another thing we would like to have is storing the equivalent, but for the very top level instance id. E.g.,:

```bash
FLUX_TOP_LEVEL_ID=xxx
```

I am adding [a command](https://github.com/converged-computing/usernetes-python/pull/3/files) `flux usernetes top-level` for our usernetes case that walks up the chain of flux parent-uri until it gets to the top level parent to interact with, and if that is orchestrated more simply via a passed envar (to avoid all the operations to get flux-uris of the parent) that might be a better solution? The use case is in prolog/epilog, for cases when we need to store metadata across all levels of some root instance, it makes sense to put at the top level for everyone to find.

I think whatever you decide to come up with will be hugely helpful, so thank you in advance!

**Comment by chu11** [comment]
> I think if FLUX_JOB_ID is set, it would return that, otherwise flux getattr jobid. If that failed, then it would return empty?

Yeah, this is what I was thinking.  Just wrap the logic into it.

**Comment by trws** [comment]
That would be fine I think, it's consistency that matters most.  Having it be a command might be best since that means things like `flux toplevel ...` would work, or if we have one for `parent` or whatever right, it's more composable that way.

I admit to a personal preference to have access to at least the innermost job ID and matching flux URI be very easy though, since that's what people are most used to from other systems and will need to do naive ports of job scripts that use the enclosing jobid and talk to the system scheduler in batch scripts.

**Comment by MrBurmark** [comment]
Having it as a environment variable is more consistent with the behavior of slurm where there is an environment varaible for the id for the allocation and for the if of the subjob inside of the allocation. It also saves me from having to do something like execve or something inside of my process.

**Comment by grondo** [comment]
Good point about the extra pain of requiring an execve (or using the Flux C API) to get the `jobid` attribute.
We do try to avoid the proliferation of environment variables (a possible anti-pattern from Slurm)

We could do something simple like `FLUX_ENCLOSING_ID`, which would contain the closest enclosing jobid. This would be the same as `FLUX_JOB_ID` for jobs run directly in a top-level instance. In an initial program (what we call the batch script or program spawned by `flux alloc`) it would be set to the `jobid` attribute. Jobs run in the instance would just inherit this variable from the environment.

This captures only the _enclosing_ jobid. If `flux alloc` or `flux batch` was run within a batch or alloc job,  then this variable would be overwritten with the new batch/alloc jobid. This satisfies the original use case in this issue, but there are also cases proposed above that also want a "top level" jobid _and_ URI.  I wonder if we should take a simple approach for now to close this issue, and open a separate issue for the more complex use case of accessing the URI and jobid at any level in a hierarchy of jobs.

**Comment by grondo** [comment]
Oh, @MrBurmark, it just occurred to me as of flux-core v0.70.0 you can request an environment variable be set containing the batch jobid using template substitution, e.g.:
```console
$ flux batch -N1 --env=FLUX_BATCH_ID={{id}} ...
ƒV2VX34zvb
```

Would result in `FLUX_BATCH_ID=ƒV2VX34zvb` set in your batch environment. Be cautious that this will be propagated to all jobs (including batch/alloc jobs) run within that batch job.

**Comment by MrBurmark** [comment]
The FLUX_ENCLOSING_ID sounds perfect. The flux batch or flux alloc with a
flux run inside is the use case I am thinking about.
